### ISSUE TITLE
Instance banning a remote user should prevent them from participating in remote versions of communities

### ISSUE THREAD

**Issue by sunaurus** [LemmyNet/lemmy#3399]
### Requirements

- [X] Is this a feature request? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a feature request? Do not put multiple feature requests in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Is your proposal related to a problem?

# Problem

When a user from instanceA is banned from instanceB, they are not prevented from making posts in instanceB communities. These posts do not show up outside instanceA, but this still means that future ban-worthy content from the banned user will be shown to all other users of instanceA, and the mod teams of instanceB will be powerless to improve the content for other users from instanceA.

Additionally, it seems that under certain conditions, these local-only posts from banned users can reach out to the wider federated network, for example, when a user from another instance searches for the post's url.

### Describe the solution you'd like.

# Solution

Instance bans would ideally work just like community bans - they would be federated and go into a new `site_person_ban` table, which could be checked everywhere that currently checks `community_person_ban`, and if a ban exists in either table, the users action would be prevented.

### Describe alternatives you've considered.

N/A

**Comment by 0xAnansi** [comment]
Couldn't this be used to DoS instances by creating a huge amount of bots with the intent to get banned, getting banned and then propagate this ban to all the instances, filling up their databases?

**Comment by lionirdeadman** [comment]
Lemmy already federates the banned list and modlog. I don't see why this would be any worse than those things federating.

**Comment by 0xAnansi** [comment]
> > Couldn't this be used to DoS instances by creating a huge amount of bots with the intent to get banned, getting banned and then propagate this ban to all the instances, filling up their databases?
> 
> It's already a thing Guess we should approach that the same way we approach spam posts etc.

wth 💀

**Comment by Nutomic** [comment]
Instance bans are federated, but its possible that they are not going through in some cases.

**Comment by sunaurus** [comment]
@Nutomic instance bans are only federeated into the modlog, they are not currently being checked as community bans are when posting or commenting. The point is that a user who is instance banned can freely post garbage in communities that are hosted in the instance that he is banned from, and this will ruin the experience for other users on his instance.

**Comment by sunaurus** [comment]
This is the relevant code (which currently doesn't do much) for when a remove ban comes in:

https://github.com/LemmyNet/lemmy/blob/main/crates/apub/src/activities/block/block_user.rs#L133

OTOH In the case of community bans, the bans are actually written into the local `community_user_ban` table (which is always checked when making posts or comments):

https://github.com/LemmyNet/lemmy/blob/main/crates/apub/src/activities/block/block_user.rs#L192

**Comment by Nutomic** [comment]
Ah you are talking about the user receives a site ban from an instance which is not their home instance. That is indeed not handled. Its a variation of https://github.com/LemmyNet/lemmy/issues/1205

**Comment by iByteABit256** [comment]
Essentially a fix to this would be iterating over the communities of instance B, and for each one to run the code from [the community ban handling](https://github.com/LemmyNet/lemmy/blob/51ccf318e8129f1bea9aa725fc2081257f583ab6/crates/apub/src/activities/block/block_user.rs#L187C1-L187C1), right?

**Comment by Nutomic** [comment]
Yes that sounds about right. So you need to change the code inside `SiteOrCommunity::Site` branches.

**Comment by Nothing4You** [comment]
additional related issue:

i've noticed that unbans and restorations can federate in some cases.
for example, if i ban a remote user, then the user gets banned and unbanned on their home instance, the user is no longer banned on my instance.
if i remove content e.g. a post in a remote community, then the community's home instance removes and restores the same post, the post will also be restored in my instance.

while this certainly isn't working as intended, i think the security impact is fairly low, as such actions coming from other instances will typically lead to defederation, and similar actions could be performed by just creating new users, but i can see this, especially for the site ban, happen easily by accident, when a user gets temp banned on their home instance at any later time for likely unrelated reasons.

**Comment by dessalines** [comment]
Seems to be a dupe of #4118 , which I have a PR out for now.

**Comment by Nothing4You** [comment]
If I understand it correctly, #4464 does not address this for site bans and communities created after the ban was issued.

**Comment by sunaurus** [comment]
@Nothing4You you are correct - I will reopen this issue for now as I am working on a PR for the fix already

**Comment by DraconicNEO** [comment]
Something this should probably also change in the modlog, is to make it so when an admin site bans a remote user it'll federate the message too as "mod banned user@instance from site voyager.lemmy.ml" instead of just saying simply banning them. Since it'll help clarify that they are banned from that site and its communities.

**Comment by Nutomic** [comment]
This is not fixed, was wrongly tagged in the commit.

**Comment by DraconicNEO** [comment]
Does this being completed mean that the hack where a user is banned from multiple communities on site ban will be removed when this is added to the release? Would probably be a good idea since it ends up spamming the modlog when banning really active users from one's site.

**Comment by Nutomic** [comment]
@DraconicNEO Yes that is removed.

https://github.com/LemmyNet/lemmy/pull/5515#discussion_r2004068653
